### ISSUE TITLE
drop-all fails when schema contains tables defined using postgresql 11 table partitioning

### ISSUE THREAD

**Issue by enda-odonohoe** [liquibase/liquibase#1212]
## Environment
**Liquibase Version**: 3.8.9

**Liquibase Integration & Version**: Dropwizard 2.0.10

**Liquibase Extension(s) & Version**: NA

**Database Vendor & Version**: Postgresql 12.2

**Operating System Type & Version**: Postgresql:12 docker image

## Description
When executing the liquibase:drop-alll , the following error occurs:

```
Caused by: liquibase.exception.DatabaseException: ERROR: cannot drop inherited constraint "XXXX_fkey" of relation "XXXX_y2020m10" [Failed SQL: (0) ALTER TABLE "public"."XXXX_y2020m10" DROP CONSTRAINT "XXXX_fkey"]
	at liquibase.executor.jvm.JdbcExecutor$ExecuteStatementCallback.doInStatement(JdbcExecutor.java:402)
	at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:59)
	at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:131)
	at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:111)
	at liquibase.database.AbstractJdbcDatabase.dropDatabaseObjects(AbstractJdbcDatabase.java:801)
	at liquibase.command.core.DropAllCommand.run(DropAllCommand.java:88)
```

## Steps To Reproduce
List the steps to reproduce the behavior.

* This is a partitioned table with a foreign key constraint like to following:

```
--liquibase formatted sql

-- changeset jh:some-change
create othertable (
id bigserial primary key,
somedata text );

create table thistable (
                                 id bigserial,
                                 otherid int references othertable(otherid) NOT NULL,
                                 thisdate timestamp with time zone NOT NULL,
                                
)  PARTITION BY RANGE (thisdate);

CREATE UNIQUE INDEX thistable_unique_idx ON thistable(id, thisdate);

CREATE TABLE thistable_y2020m6 PARTITION OF thistable FOR VALUES FROM ('2020-06-01 00:00:00')  TO ('2020-06-30 23:59:59+00');
CREATE TABLE thistable_y2020m7 PARTITION OF thistable FOR VALUES FROM ('2020-07-01 00:00:00')  TO ('2020-07-31 23:59:59+00');

```

**Comment by molivasdat** [comment]
Hi @enda-odonohoe Thanks for writing this one up.  We will add it to our queue of issues to fix.

**Comment by enda-odonohoe** [comment]
FYI: Noticed two things here.  
1) As of Postgres JDBC Driver version 42.2.11 , DatabaseMetadata.getTables(s, s1, s2, new String[]{"TABLE"}) does not return PARTITIONED TABLES, so the liquibase drop-all doesn't see them and they never get dropped.
2) Adding  typesToInclude.remove(ForeignKey.class) to the dropDatabaseObjects appears to fix the error, since the drop table...cascade looks like it is removing the constraints.

**Comment by stevesaliman** [comment]
I am also running into this issue.  Looking into it further, I have a few things to add to @enda-odonohoe's observations, which might be helpful:

1. Adding `typesToInclude.remove(ForeignKey.class)` to `AbstractJdbcDatabase.dropDatabaseObjects` will cause H2 to break.  However, this method already removes foreign keys from the types to include when the database supports disabling foreign keys.  Adding `getShortName().equals("postgresql") to the condition will take care of Postgres without breaking H2 (line 757).  I don't know if this would be needed once the JDBC driver is fixed, but my hunch is that we would still needed because I think the foreign key constraints are dropped before the tables.
2. Adding this code will allow the `dropAll` command to run, but the partitioned table won't be dropped until Postgres fixes the JDBC driver - it should also be a one line fix, but I don't know the ramifications of adding that line.
3. Because `dropAll` won't remove the table, environments like CI/CD environments that depend on being able to do a `dropAll` followed by `update` will have problems because the table already exists.  The workaround is to add the following changeset (groovy example given) before the one that creates the table:
```groovy
    changeSet(id: '1.0.0-pre-drop-my_table', author: 'Steve Saliman') {
        preConditions(onFail: 'MARK_RAN') {
            sqlCheck(expectedResult: '1') {
                "select count(*) from pg_catalog.pg_class where relname = 'my_table';"
            }
        }
        dropTable(tableName: 'my_table')
    }
```
There is a `tableExists` precondition, but it doesn't work, probably due to the same JDBC driver bug, which is why I'm using a `sqlCheck`.

**Comment by stevesaliman** [comment]
Looking into it further, I don't think this is an issue with the Postgres JDBC driver.  It supports partitioned tables, but it uses a different type for them, Liquibase isn't using that type when it calls `DatabaseMetadata.getTables`.

The fix to this needs to be in `JdbcSnapshot.getTables()`.  It has a `ResultSetCache.SingleResultSetExtractor` that declares a `ResultSetCache.SingleResultSetExtractor()` method, which needs to to add "PARTITIONED TABLE" to the list of types it passed to the JDBC driver when we're in Postgres.  Doing this will fix the problem, but I'm not certain what other side effects this change would have - I suspect they would be good ones.  Also, as I suspected, we need to remove Foreign Keys from the types to include for this to all work.

I'll submit a patch, and the good folks at Liquibase, who know the code base better than I do, can make sure there isn't anything bad about the change.

**Comment by cyraid** [comment]
I'm testing with liquibase and it seems to not even capture the partition tables in a diff. Is this a bug / missing feature?

**Comment by molivasdat** [comment]
Hi @cyraid Which version of Liquibase are you using to test?
If there is a problem with diff then I would create a new issue.  We fixed dropAll in 4.3.x https://github.com/liquibase/liquibase/pull/1650

**Comment by cyraid** [comment]
Hi @molivasdat , I guess I'm not just seeing the partitions in the diff. But I did create a new issue. #1874

**Comment by nvoxland** [comment]
I'll close this issue as resolved, since #1650 solved this in 4.3, and there is the separate issue for including them in the diff.
